File: src/syncai_backend/syncai_backend/interfaces/rest/routers/schedule.py
import structlog
from datetime import datetime
from typing import List, Optional
from fastapi import APIRouter
from pydantic import BaseModel, Field, model_validator

from syncai_backend.gateways.workflow.schema import (
    Step,
    ScheduleTask,
    ScheduleTrigger,
    WorkflowTaskDefinition,
)
from syncai_backend.gateways.workflow.workflow import WorkflowGateway
from syncai_backend.interfaces.rest.routers.task import StepRequest


class ScheduleTriggerRequest(BaseModel):
    cron: Optional[str] = Field(
        default=None,
        description="Cron expression. Provide exactly one of cron or interval_seconds.",
        examples=["*/3 * * * *"],
    )
    interval_seconds: Optional[int] = Field(
        default=None,
        gt=0,
        description="Fire every N seconds. Provide exactly one of cron or interval_seconds.",
        examples=[1800],
    )
    timezone: Optional[str] = Field(
        default=None,
        description="IANA timezone applied to cron (ignored for interval).",
        examples=["Asia/Taipei"],
    )

    @model_validator(mode="after")
    def _exactly_one(self) -> "ScheduleTriggerRequest":
        if bool(self.cron) == bool(self.interval_seconds):
            raise ValueError("Provide exactly one of 'cron' or 'interval_seconds'")
        return self


class ScheduleRequest(BaseModel):
    id: str = Field(
        ...,
        description="Unique identifier of the schedule",
        examples=["robot01-daily-move"],
    )
    trigger: ScheduleTriggerRequest = Field(..., description="When the schedule fires")
    steps: List[StepRequest] = Field(
        ..., description="Steps executed on each trigger", examples=[]
    )
    map_name: Optional[str] = Field(
        default=None,
        description=(
            "Map whose frame the MOVE coordinates are in, carried through to the "
            "schedule memo so a client can tell whether this schedule still "
            "belongs to the loaded map. A display label only, and deliberately "
            "unvalidated: this router has no MapCatalogRepo, and threading one in "
            "for a label is not worth it. POST /api/v1/saved_tasks/{id}/schedule "
            "is the validated path."
        ),
    )


class ScheduleResponse(BaseModel):
    id: str = Field(..., description="Unique identifier of the schedule")
    message: str = Field(..., description="Additional information about the schedule")


class ScheduleTriggerResponse(BaseModel):
    cron: Optional[str] = None
    interval_seconds: Optional[int] = None
    timezone: Optional[str] = None


class ScheduleStateResponse(BaseModel):
    id: str = Field(..., description="Unique identifier of the schedule")
    trigger: ScheduleTriggerResponse = Field(..., description="When the schedule fires")
    paused: bool = Field(..., description="Whether the schedule is paused")
    next_run_times: List[datetime] = Field(
        ..., description="Upcoming trigger times (UTC)"
    )
    map_name: Optional[str] = Field(
        default=None,
        description="Map whose frame this schedule's MOVE coordinates are in.",
    )
    saved_task_id: Optional[str] = Field(
        default=None,
        description="The saved task this schedule was frozen from, if any.",
    )
    saved_task_name: Optional[str] = Field(
        default=None, description="That saved task's name at registration time."
    )
    # The element is StepRequest rather than a new response model: it is already
    # exactly {id, type, params}, it is already imported here, and a second
    # identical model is a copy that can drift. (StepState is the wrong shape --
    # that one is status/error, not a definition.)
    steps: List[StepRequest] = Field(
        default_factory=list,
        description=(
            "The frozen step list. Populated by GET /api/v1/schedules/{id} only; "
            "the collection endpoint always answers [], because Temporal's "
            "schedule *list* API does not carry the start-workflow arguments. "
            "Frozen at registration: later vertex edits do not reach a scheduled "
            "run, so a client comparing these against their source saved task is "
            "how staleness becomes visible."
        ),
    )


def init_schedule_router(
    logger: structlog.stdlib.BoundLogger, workflow_gw: WorkflowGateway
) -> APIRouter:
    schedule_router = APIRouter(prefix="", tags=["Schedule"])

    @schedule_router.post("/api/v1/schedules", response_model=ScheduleResponse)
    async def create_schedule(req: ScheduleRequest):
        schedule_task = ScheduleTask(
            id=req.id,
            trigger=ScheduleTrigger(
                cron=req.trigger.cron,
                interval_seconds=req.trigger.interval_seconds,
                timezone=req.trigger.timezone,
            ),
            definition=WorkflowTaskDefinition(
                steps=[
                    Step(id=step.id, type=step.type, params=step.params)
                    for step in req.steps
                ],
            ),
            map_name=req.map_name,
        )

        await workflow_gw.create_schedule(schedule=schedule_task)

        return ScheduleResponse(
            id=req.id,
            message=f"Schedule {req.id} has been created.",
        )

    @schedule_router.get(
        "/api/v1/schedules", response_model=List[ScheduleStateResponse]
    )
    async def list_schedules():
        views = await workflow_gw.list_schedules()

        return [
            ScheduleStateResponse(
                id=view.id,
                trigger=ScheduleTriggerResponse(
                    cron=view.trigger.cron,
                    interval_seconds=view.trigger.interval_seconds,
                    timezone=view.trigger.timezone,
                ),
                paused=view.paused,
                next_run_times=view.next_run_times,
                map_name=view.map_name,
                saved_task_id=view.saved_task_id,
                saved_task_name=view.saved_task_name,
                # `steps` left at its default here — see the field's description.
            )
            for view in views
        ]

    @schedule_router.get(
        "/api/v1/schedules/{id}",
        response_model=ScheduleStateResponse,
        # The only route here that serialises StepParams outward, and FastAPI
        # defaults this flag to True: MoveParams / ArtifactParams inherit
        # BaseSchema's to_camel generator, so the default would answer
        # `artifactId` / `waitFor` / `waitTimeoutSeconds` while the request side
        # accepts both spellings — a silent asymmetry at a boundary the frontend
        # documents as snake_case-only.
        response_model_by_alias=False,
    )
    async def get_schedule(id: str):
        view = await workflow_gw.get_schedule(schedule_id=id)

        return ScheduleStateResponse(
            id=view.id,
            trigger=ScheduleTriggerResponse(
                cron=view.trigger.cron,
                interval_seconds=view.trigger.interval_seconds,
                timezone=view.trigger.timezone,
            ),
            paused=view.paused,
            next_run_times=view.next_run_times,
            map_name=view.map_name,
            saved_task_id=view.saved_task_id,
            saved_task_name=view.saved_task_name,
            steps=[
                StepRequest(id=step.id, type=step.type, params=step.params)
                for step in view.steps
            ],
        )

    @schedule_router.delete("/api/v1/schedules/{id}", response_model=ScheduleResponse)
    async def delete_schedule(id: str):
        await workflow_gw.delete_schedule(schedule_id=id)

        return ScheduleResponse(
            id=id,
            message=f"Schedule {id} has been deleted.",
        )

    @schedule_router.post(
        "/api/v1/schedules/{id}/pause", response_model=ScheduleResponse
    )
    async def pause_schedule(id: str):
        await workflow_gw.pause_schedule(schedule_id=id)

        return ScheduleResponse(
            id=id,
            message=f"Schedule {id} has been paused.",
        )

    @schedule_router.post(
        "/api/v1/schedules/{id}/resume", response_model=ScheduleResponse
    )
    async def resume_schedule(id: str):
        await workflow_gw.resume_schedule(schedule_id=id)

        return ScheduleResponse(
            id=id,
            message=f"Schedule {id} has been resumed.",
        )

    return schedule_router

File: src/syncai_frontend/lib/api/schedule.ts
// Client for the backend schedule API
// (src/syncai_backend/syncai_backend/interfaces/rest/routers/schedule.py).
//
// Separate from lib/api/task.ts even though a schedule's `steps` are literally
// the same `StepRequest`, because the two have opposite lifecycles. A task is
// created and immediately dispatched — creation *is* dispatch — and then polled
// once until it is terminal, after which it is gone. A schedule is a durable
// registration that is listed, paused, resumed and deleted, and never polled.
// Merging them would leave one file whose header has to explain both.
//
// snake_case throughout, and `interval_seconds` in particular. `ScheduleRequest`
// and `ScheduleTriggerRequest` are plain pydantic BaseModels with no alias
// generator, unlike the `BaseSchema` in gateways/workflow/schema.py that the step
// params inherit from. So `intervalSeconds` is not an alias here: it is silently
// ignored, `interval_seconds` stays None, and the request then fails the
// `_exactly_one` validator with a 422 naming a field the client never sent.

import { apiUrl } from "@/lib/api/config";
import { requestJson } from "@/lib/api/http";
import type { TaskStepRequest } from "@/lib/api/task";

/**
 * `ScheduleTriggerRequest` / `ScheduleTriggerResponse`.
 *
 * Exactly one of `cron` / `interval_seconds` — the backend's validator rejects
 * both and neither with equal prejudice, so the two are modelled as optional
 * rather than as a union: the form that builds one only ever sets a single field.
 *
 * Every field is `| null` because the *response* is not the request shape read
 * back. `ScheduleTriggerResponse` declares all three as `Optional[...] = None`
 * and FastAPI serialises them, so an interval schedule answers with an explicit
 * `{"cron": null, "interval_seconds": 86400, "timezone": null}` rather than
 * omitting the two that do not apply. Consumers must test truthiness, not
 * `=== undefined`.
 */
export interface ScheduleTrigger {
  cron?: string | null;
  /** Must be > 0. Cannot be combined with `cron`. */
  interval_seconds?: number | null;
  /**
   * IANA zone name. Only meaningful for `cron` — the interval path ignores it,
   * so it is omitted rather than sent-and-ignored (see schedule-form.tsx).
   */
  timezone?: string | null;
}

/** What the operator supplies. `steps` is the same wire shape a task takes. */
export interface ScheduleDraft {
  id: string;
  trigger: ScheduleTrigger;
  steps: readonly TaskStepRequest[];
}

/** `ScheduleStateResponse`, verbatim. */
export interface ScheduleState {
  id: string;
  /**
   * The trigger as it was registered, read back from the Temporal schedule's
   * memo rather than from the schedule spec — Temporal normalises a cron string
   * into calendar fields, so the spec can no longer say what was asked for.
   */
  trigger: ScheduleTrigger;
  paused: boolean;
  /** ISO-8601 **UTC**, soonest first. Empty while the schedule is paused. */
  next_run_times: string[];
  /** Map the frozen coordinates are in, from the memo. Null for older schedules. */
  map_name?: string | null;
  /** The saved task this was frozen from, if it came from one. */
  saved_task_id?: string | null;
  saved_task_name?: string | null;
  /**
   * The frozen step list — **only** populated by `getSchedule`; always `[]` from
   * `listSchedules`, and legitimately `[]` for a schedule registered before this
   * field existed or one whose payload could not be decoded.
   *
   * Frozen at registration: a schedule stores concrete steps in Temporal, and
   * nothing re-reads them, so later vertex edits reach saved tasks and immediate
   * dispatches but not an already-registered schedule. Comparing these against
   * the source saved task's current resolution is how that staleness becomes
   * visible instead of silent.
   */
  steps?: TaskStepRequest[];
}

/**
 * The operator types the id, so `encodeURIComponent` is load-bearing rather than
 * defensive here: an unescaped `/` or `#` in a name would rewrite the URL into a
 * different route instead of producing a 404. Same reason as `vertexPath`.
 */
function schedulePath(id?: string, action?: "pause" | "resume"): string {
  const base = "/api/v1/schedules";
  if (!id) return apiUrl(base);
  const one = `${base}/${encodeURIComponent(id)}`;
  return apiUrl(action ? `${one}/${action}` : one);
}

/**
 * Every registered schedule for this robot — but never their steps.
 *
 * `steps` is always `[]` from this endpoint, and that is a property of Temporal
 * rather than an omission: a schedule *list* element carries only the workflow
 * type name, so reaching the frozen step list would cost a `describe` RPC per
 * row, paid on first paint for data most rows never show. `getSchedule` is the
 * path that carries them.
 */
export function listSchedules(signal?: AbortSignal): Promise<ScheduleState[]> {
  return requestJson<ScheduleState[]>(schedulePath(), { signal });
}

/**
 * One schedule, **including its frozen step list**.
 *
 * This is the only way to see what a schedule will actually do. Fetched on a
 * deliberate gesture — expanding the row — so the extra describe is paid once,
 * for the one schedule the operator asked about.
 */
export function getSchedule(
  id: string,
  signal?: AbortSignal,
): Promise<ScheduleState> {
  return requestJson<ScheduleState>(schedulePath(id), { signal });
}

// The four writes all drop the `{id, message}` envelope: it says nothing the
// caller does not already know from the request having succeeded, which is the
// same reasoning `deleteVertex` records. What a caller *does* need afterwards —
// the recomputed `next_run_times` — is only knowable by asking, so every one of
// these is followed by a list refresh rather than a local splice.

export function createSchedule(draft: ScheduleDraft): Promise<void> {
  return requestJson<void>(schedulePath(), {
    method: "POST",
    body: JSON.stringify(draft),
    parse: false,
  });
}

export function pauseSchedule(id: string): Promise<void> {
  return requestJson<void>(schedulePath(id, "pause"), {
    method: "POST",
    parse: false,
  });
}

export function resumeSchedule(id: string): Promise<void> {
  return requestJson<void>(schedulePath(id, "resume"), {
    method: "POST",
    parse: false,
  });
}

export function deleteSchedule(id: string): Promise<void> {
  return requestJson<void>(schedulePath(id), {
    method: "DELETE",
    parse: false,
  });
}
